### PR TITLE
[cli] Expand bbrc related options in the parser

### PR DESCRIPTION
**Pull request by maggie-lou**
After this PR, bbrc related options will be parsed for Bazel commands.
Ex:
``` .bbrc
run --stream_run_logs
```

A future PR in the stack will parse bbrc options for bb CLI commands (like bb remote)

### CODE REVIEW

File: cli/arg/arg.go [hunk 1]
```diff
@@ -8,26 +8,38 @@ import (
...
+	// All bazelrc files, --config flags, bbrc files, and --bb_config flags are expanded.
 	resolved *parsed.OrderedArgs
 
 	// noResolve indicates that resolved args (i.e. --config and --bazelrc flags expanded) should not be computed.
 	noResolve bool
 }
 
-// Forwarded returns the forwarded args as a canonicalized []string.
+// Forwarded returns the args that should be passed to Bazelisk as a canonicalized []string.
 func (a *BazelArgs) Forwarded() []string {
```
[the remark this answers is not in the document]
Comment (maggie-lou) [reply cli/arg/arg.go:34]: Fixed edge cases around the watcher, sidecar, and plugins, and added integration tests

File: cli/parser/parser.go [hunk 2]
```diff
@@ -927,10 +946,30 @@ func (p *Parser) ParseBBRCFiles(workspaceDir string, filePaths ...string) (map[s
 			_, isBBCommand := cli_command.CommandsByName[phase]
 			return bazelrc.IsPhase(phase) || isBBCommand
 		},
-		ParsePhase: p.ParseConfig,
+		ParsePhase: p.parseBBRCConfig,
 	}, filePaths...)
 }
 
+func (p *Parser) parseBBRCConfig(phase string, tokens []string) ([]arguments.Argument, error) {
```
[the remark this answers is not in the document]
Comment (maggie-lou) [reply cli/parser/parser.go:980]: Fixed and added a test

File: cli/parser/parser.go [hunk 3]
```diff
@@ -840,16 +833,42 @@ func (p *Parser) ResolveArgs(parsedArgs *parsed.OrderedArgs) (*parsed.OrderedArg
...
+
 func (p *Parser) resolveArgs(parsedArgs *parsed.OrderedArgs, ws string) (*parsed.OrderedArgs, error) {
-	// TODO(Maggie): Add bbrc config expansion here
+	// When expanding bbrc, use the bb-specific parser because bbrc files
+	// should not contain Bazel options.
+	bbParser, err := GetBBParserForCommand(parsedArgs.GetCommand())
+	if err != nil {
+		return nil, err
+	}
+	parsedArgs, err = bbParser.expandBBRC(parsedArgs, ws)
```
[the remark this answers is not in the document]
Comment (maggie-lou) [reply cli/parser/parser.go:853]: Fixed. `help` will no longer expand bbrc flags. Added integration tests to ensure `help` still correctly expands bazelrc flags

File: cli/parser/parser.go [hunk 4]
```diff
@@ -840,16 +833,42 @@ func (p *Parser) ResolveArgs(parsedArgs *parsed.OrderedArgs) (*parsed.OrderedArg
...
 	configs, defaultConfig, err := p.consumeAndParseRCFiles(parsedArgs, ws)
 	if err != nil {
 		return nil, err
 	}
 	return bazelrc.ExpandConfigs(parsedArgs, configs, defaultConfig)
 }
 
-func (p *Parser) expandBBRC(args *parsed.OrderedArgs, workspaceDir, homeDir string) (*parsed.OrderedArgs, error) {
+func (p *Parser) expandBBRC(args *parsed.OrderedArgs, workspaceDir string) (*parsed.OrderedArgs, error) {
+	homeDir, err := os.UserHomeDir()
```
[the remark this answers is not in the document]
Comment (maggie-lou) [reply cli/parser/parser.go]: Fixed

File: cli/cmd/bb/bb.go [hunk 5]
```diff
@@ -289,7 +289,9 @@ func runHelp(args *parsed.OrderedArgs) (int, error) {
 // EXPLICIT_COMMAND_LINE metadata to the bazel invocation.
 func handleBazelCommand(start time.Time, bazelArgs *arg.BazelArgs, execArgs []string, originalArgs []string) (exitCode int, err error) {
 	// Maybe run interactively (watching for changes to files).
-	if exitCode, err := watcher.Watch(append([]string{os.Args[0]}, arg.JoinExecutableArgs(bazelArgs.Forwarded(), execArgs)...)); exitCode >= 0 || err != nil {
+	// The watcher starts another bb process, so retain the complete set of original arguments.
+	// They'll be reparsed by the new process.
+	if exitCode, err := watcher.Watch(append([]string{os.Args[0]}, arg.JoinExecutableArgs(bazelArgs.Unresolved(), execArgs)...)); exitCode >= 0 || err != nil {
```
[the remark this answers is not in the document]
Comment (maggie-lou) [reply cli/cmd/bb/bb.go:294]: Fixed and added a test

File: cli/parser/parser.go [hunk 6]
```diff
@@ -840,16 +834,42 @@ func (p *Parser) ResolveArgs(parsedArgs *parsed.OrderedArgs) (*parsed.OrderedArg
...
 	configs, defaultConfig, err := p.consumeAndParseRCFiles(parsedArgs, ws)
 	if err != nil {
 		return nil, err
 	}
 	return bazelrc.ExpandConfigs(parsedArgs, configs, defaultConfig)
 }
 
-func (p *Parser) expandBBRC(args *parsed.OrderedArgs, workspaceDir, homeDir string) (*parsed.OrderedArgs, error) {
+func (p *Parser) expandBBRC(args *parsed.OrderedArgs, workspaceDir string) (*parsed.OrderedArgs, error) {
+	homeDir, err := userHomeDir()
```
[the remark this answers is not in the document]
Comment (maggie-lou) [reply cli/parser/parser.go:869]: Fixed

File: cli/parser/bbrc/bbrc.go [hunk 7]
```diff
@@ -62,3 +62,9 @@ func ExpandConfigs(
 ) (*parsed.OrderedArgs, error) {
 	return args.ExpandConfigsWithPolicy(namedConfigs, defaultConfig, NewConfigExpansionPolicy())
 }
+
+// RemoveOptions removes all options related to .bbrc files.
+func RemoveOptions(args *parsed.OrderedArgs) {
```
[the remark this answers is not in the document]
Comment (maggie-lou) [reply cli/parser/bbrc/bbrc.go:67]: Not addressing here

File: cli/parser/parser.go [hunk 8]
```diff
@@ -927,10 +959,44 @@ func (p *Parser) ParseBBRCFiles(workspaceDir string, filePaths ...string) (map[s
...
+	}
+	// Validate that all options in the config are CLI flags.
+	// Bazel options are not allowed in bbrc files.
+	for _, classified := range parsed.Classify(args) {
+		option, ok := classified.(parsed.ClassifiedOption)
+		if !ok {
+			continue
+		}
+		switch option.AsOption().PluginID() {
+		case options.UnknownBuiltinPluginID, options.StarlarkBuiltinPluginID:
```
Comment (tempoz) [inline cli/parser/parser.go]:
```suggestion
		if id := option.AsOption().PluginID(); id == options.UnknownBuiltinPluginID || id == options.StarlarkBuiltinPluginID {
```
There's nothing semantically wrong here, but using a switch statement for a single case is pretty unusual.

**Review by tempoz — APPROVED** [review]

File: cli/parser/parser.go (hunk 8 above)
Comment (maggie-lou) [reply cli/parser/parser.go]: Fixed
